nixos/modules/services/networking/knot.nix

```diff
@@ -210,19 +270,35 @@ in {
       wants = [ "network.target" ];
       after = ["network.target" ];
 
-      serviceConfig = {
+      serviceConfig = let
+        # https://www.knot-dns.cz/docs/3.3/singlehtml/index.html#pre-requisites
+        xdpCapabilities = lib.optionals (cfg.enableXDP) [
+          "CAP_NET_ADMIN"
+          "CAP_NET_RAW"
+          "CAP_SYS_ADMIN"
+          "CAP_IPC_LOCK"
+        ] ++ lib.optionals (lib.versionOlder config.boot.kernelPackages.kernel.version "5.11") [
+          "CAP_SYS_RESOURCE"
```

Comment:
This could theoretically be tricky for people who do `nixos-rebuild switch` when kernel branch changes (instead of `nixos-rebuild boot`).  But it seems OK not to safeguard against that here, as it's quite a rare combination in this case: knot+XDP + _downgrading_ across 5.11 boundary.